### ISSUE TITLE
Beatmap Info Wedge Map Author Link is Incorrect

### ISSUE THREAD

**Issue by LeNitrous** [ppy/osu#12965]
**Describe the bug:** The beatmap info wedge map author when clicked sends you to the wrong profile.

This is due to recent API changes that affected `BeatmapMetadata` which is done in this pull request: https://github.com/ppy/osu-web/pull/7517. The `user_id` property now uses the difficulty owner set via beatmap discussions but the `creator` property still uses the beatmap set host.

A suggestion which may be worthwhile is instead of showing `mapped by <creator>`, it should instead show `hosted by <host> and mapped by <creator>`. However, the API endpoint responsible for fetching the beatmap metadata only returns the difficulty owner's ID and requires another lookup in the API.

To reproduce:
- In beatmap discussions, change the most **bottom-most** (most difficult) difficulty owner to anyone else
![image](https://user-images.githubusercontent.com/20495991/119677382-413c8800-be71-11eb-9a0b-3e30e362688b.png)

- If the beatmap already exists in lazer, delete it and restart the game.
- Download the game via the web, in-game, or import via stable.

I'll attempt at fixing this if I can be assigned in this issue.

**Screenshots or videos showing encountered issue:** 
![ymk4jfaKjo](https://user-images.githubusercontent.com/20495991/119677619-75b04400-be71-11eb-868a-e2a183e5227f.gif)

**osu!lazer version:**  master

**Logs:** N/A

**Comment by peppy** [comment]
This is wrong in multiple places, so keep that in mind when looking into a fix (needs to be fixed in the local lookup too).

**Comment by frenzibyte** [comment]
Seeing various reports on this across the years, I think it's good to prioritise it. Shouldn't take much effort anyway.

**Comment by bdach** [comment]
Another issue that stems from incorrect client-side handling of guest mappers is that in some cases it stops the "My Maps" grouping mode from working correctly. One reported case is https://osu.ppy.sh/community/forums/topics/2139456?n=1.

The database contains this garbage:

<img width="548" height="150" alt="Image" src="https://github.com/user-attachments/assets/53120f4c-160c-4ad1-b0a6-4b01ce894a1b" />

where 2 of the 4 difficulties have an invalid username for the mapper IDs. This then fails later at

https://github.com/ppy/osu/blob/03455ba683f35228fc4a58120892675922f52fc6/osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs#L424-L427

because `beatmap.BeatmapSet!.Metadata.Author` will arbitrarily pick one of the difficulties' metadata at random, meaning that if a guest difficulty gets picked there, the mapset won't be recognized as the local user's.

**Comment by zihadmahiuddin** [comment]
I see that the issue is assigned but I was doing some experiment for fun and looks like the following diff is enough to fix the issue for new beatmaps (although it might be better to make `AuthorName` non-nullable and have osu-web send the name along with the ID, separate from `owners`):

```diff
diff --git a/osu.Game/Beatmaps/APIBeatmapMetadataSource.cs b/osu.Game/Beatmaps/APIBeatmapMetadataSource.cs
index e15de7ec02..9865e9b4c2 100644
--- a/osu.Game/Beatmaps/APIBeatmapMetadataSource.cs
+++ b/osu.Game/Beatmaps/APIBeatmapMetadataSource.cs
@@ -3,6 +3,7 @@
 
 using System;
 using System.Diagnostics;
+using System.Linq;
 using osu.Game.Database;
 using osu.Game.Online.API;
 using osu.Game.Online.API.Requests;
@@ -58,6 +59,7 @@ public bool TryLookup(BeatmapInfo beatmapInfo, out OnlineBeatmapMetadata? online
                         BeatmapID = res.OnlineID,
                         BeatmapSetID = res.OnlineBeatmapSetID,
                         AuthorID = res.AuthorID,
+                        AuthorName = res.BeatmapOwners.FirstOrDefault(o => o.Id == res.AuthorID)?.Username,
                         BeatmapStatus = res.Status,
                         BeatmapSetStatus = res.BeatmapSet?.Status,
                         DateRanked = res.BeatmapSet?.Ranked,
diff --git a/osu.Game/Beatmaps/BeatmapUpdaterMetadataLookup.cs b/osu.Game/Beatmaps/BeatmapUpdaterMetadataLookup.cs
index d2d9a54fba..94de3df77a 100644
--- a/osu.Game/Beatmaps/BeatmapUpdaterMetadataLookup.cs
+++ b/osu.Game/Beatmaps/BeatmapUpdaterMetadataLookup.cs
@@ -79,6 +79,8 @@ public void Update(BeatmapSetInfo beatmapSet, bool preferOnlineFetch)
                     beatmapInfo.Metadata.Author.OnlineID = res.AuthorID;
                     beatmapInfo.Metadata.UserTags.Clear();
                     beatmapInfo.Metadata.UserTags.AddRange(res.UserTags);
+
+                    if (res.AuthorName != null) beatmapInfo.Metadata.Author.Username = res.AuthorName;
                 }
             }
 
diff --git a/osu.Game/Beatmaps/OnlineBeatmapMetadata.cs b/osu.Game/Beatmaps/OnlineBeatmapMetadata.cs
index 2acadde352..75b61e0003 100644
--- a/osu.Game/Beatmaps/OnlineBeatmapMetadata.cs
+++ b/osu.Game/Beatmaps/OnlineBeatmapMetadata.cs
@@ -29,6 +29,11 @@ public class OnlineBeatmapMetadata
         /// </summary>
         public int AuthorID { get; init; }
 
+        /// <summary>
+        /// The name of the author.
+        /// </summary>
+        public string? AuthorName { get; init; }
+
         /// <summary>
         /// The online status of the beatmap.
         /// </summary>
diff --git a/osu.Game/Screens/Select/BeatmapCarouselFilterGrouping.cs b/osu.Game/Screens/Select/BeatmapCarouselFilterGrouping.cs
index cfbade0bde..8805da38aa 100644
--- a/osu.Game/Screens/Select/BeatmapCarouselFilterGrouping.cs
+++ b/osu.Game/Screens/Select/BeatmapCarouselFilterGrouping.cs
@@ -467,7 +467,9 @@ private List<GroupMapping> defineGroupsByCollection(List<CarouselItem> carouselI
 
         private IEnumerable<GroupDefinition> defineGroupByOwnMaps(BeatmapInfo beatmap, int? localUserId, string? localUserUsername)
         {
-            var author = beatmap.BeatmapSet!.Metadata.Author;
+            var beatmapMetadata = beatmap.BeatmapSet!.Beatmaps.FirstOrDefault(b => b.OnlineID == beatmap.OnlineID)?.Metadata ?? new BeatmapMetadata();
+            var author = beatmapMetadata.Author;
 
             if (author.OnlineID == localUserId || (author.OnlineID <= 1 && author.Username == localUserUsername))
                 return new GroupDefinition(0, "My maps").Yield();

```

But for beatmaps that were previously imported, it stores the incorrect username (parsed from `.osu`) with the correct user ID (from `GetBeatmapRequest`).

> needs to be fixed in the local lookup too

Do you mean `LocalCachedBeatmapMetadataSource` here? Because `online.db` only contains the user ID too, username is missing there, so that can't be fixed without changing the database schema I guess.

**Comment by bdach** [comment]
@zihadmahiuddin issue is assigned for a reason, your patch is fundamentally a non-starter because it's still assuming there is only one author of a beatmap. please find another, non-assigned issue to work on

**Comment by bdach** [comment]
Given that solving this issue is a long-standing goal of mine and I am currently a bit in review debt I decided to revisit this and see if I could do some R&D at seeing what the road forward to properly resolving this would be here. To that end, the following branch was created: <https://github.com/ppy/osu/compare/master...bdach:osu:guest-difficulties?expand=1>. Be aware that this branch is a draft of a draft and has many issues. It is mostly just a test of the approach I took, but when used *just right* it can be finagled to work correctly:

<img width="2560" height="965" alt="Image" src="https://github.com/user-attachments/assets/ae7e5a6c-2f59-4350-ada6-4c408a500957" />

The base idea here is to eviscerate `BeatmapMetadata.Author` and replace it with two complementary properties: `BeatmapSet.Host` and `BeatmapInfo.Authors`.

The reason why both have to exist is that both are used in various places. For instance, stringifying a set for display requires the host, while stringifying *a difficulty* requires the authors / guest mappers of that particular difficulty.

On the way beelining to the above result the following (relatively) minor problems surfaced that would need to be addressed for a final go at this:

- `online.db` does not contain usable data for determining who is the host of a set. In https://github.com/ppy/osu-onlinedb-generator/pull/12 I added the guest mapper data, but I did not foresee that the host info would be required as well. The host ID has to make it to `online.db` for this to be viable (doing so will also make backwards population much less painful and much more correct out of the box).
- The editor is a weird pain point here because of decoding. The `Creator:` key in the `.osu` is at this point mostly useless and misleading in a world rife with guest difficulties. The ranking criteria - stuck between a rock and a hard place - has enforced that all difficulties of a beatmap need to specify the *host*'s username in the `.osu`. The branch I linked probably does not handle that correctly at all. The manipulations at hand are a little tricky and would need to be well tested.
- It's a little unclear what to do with song select grouping & sorting when this duality is instituted. The way I wrote this was the way I found most intuitive, which is that "Author" sorting & grouping mode does not split the difficulties apart and uses the host's username (this may seem arbitrary, but what happens when a difficulty has multiple authors?). However, "My Maps" grouping *will* take guest participation into account because that's a binary choice between showing and not showing the map and that's a much easier ask.

Given that the *entry point* here is an 83 files changed diff I will not attempt to continue this short term, but hope to revisit this eventually if review availability becomes less scarce.
